src/core/lib/event_engine/windows/iocp.cc

```diff
@@ -0,0 +1,146 @@
+// Copyright 2022 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <grpc/support/port_platform.h>
+
+#ifdef GPR_WINDOWS
+#include <chrono>
+
+#include "absl/strings/str_format.h"
+
+#include <grpc/support/alloc.h>
+#include <grpc/support/log_windows.h>
+
+#include "src/core/lib/event_engine/trace.h"
+#include "src/core/lib/event_engine/windows/iocp.h"
+#include "src/core/lib/event_engine/windows/socket.h"
+
+namespace grpc_event_engine {
+namespace experimental {
+
+IOCP::IOCP(EventEngine* event_engine) noexcept
+    : event_engine_(event_engine),
+      iocp_handle_(CreateIoCompletionPort(INVALID_HANDLE_VALUE, NULL,
+                                          (ULONG_PTR)NULL, 0)) {
+  GPR_ASSERT(iocp_handle_);
+  WSASocketFlagsInit();
+}
+
+// Shutdown must be called prior to deletion
+IOCP::~IOCP() {}
+
+WrappedSocket* IOCP::Watch(SOCKET socket) {
+  WinWrappedSocket* wrapped_socket =
+      new WinWrappedSocket(socket, event_engine_);
+  HANDLE ret =
+      CreateIoCompletionPort(reinterpret_cast<HANDLE>(socket), iocp_handle_,
+                             reinterpret_cast<uintptr_t>(wrapped_socket), 0);
+  if (!ret) {
+    char* utf8_message = gpr_format_message(WSAGetLastError());
+    gpr_log(GPR_ERROR, "Unable to add socket to iocp: %s", utf8_message);
+    gpr_free(utf8_message);
+    __debugbreak();
+    abort();
+  }
+  GPR_ASSERT(ret == iocp_handle_);
+  return wrapped_socket;
+}
+
+void IOCP::Shutdown() {
+  while (outstanding_kicks_.load() > 0) {
+    Work(std::chrono::hours(42));
+  }
+  GPR_ASSERT(CloseHandle(iocp_handle_));
+}
+
+absl::Status IOCP::Work(EventEngine::Duration timeout) {
+  DWORD bytes = 0;
+  ULONG_PTR completion_key;
+  LPOVERLAPPED overlapped;
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_event_engine_trace)) {
+    gpr_log(GPR_DEBUG, "IOCP::%p doing work", this);
+  }
+  BOOL success = GetQueuedCompletionStatus(
+      iocp_handle_, &bytes, &completion_key, &overlapped,
+      static_cast<DWORD>(
+          std::chrono::duration_cast<std::chrono::milliseconds>(timeout)
+              .count()));
+  if (success == 0 && overlapped == NULL) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_event_engine_trace)) {
+      gpr_log(GPR_DEBUG, "IOCP::%p deadline exceeded", this);
+    }
+    return absl::DeadlineExceededError(
```

Comment:
note that deadline exceeded will eventually happen for every timer expiry, so we probably don't want to be allocating here

Reply:
Good catch. This now returns a (copy of a) static const status.